### ISSUE TITLE
/v1/agent/self endpoint returning incorrect config values

### ISSUE THREAD

**Issue by watsonian** [hashicorp/consul#8728]
#### Overview of the Issue

The `/v1/agent/self` is described as "return[ing] the configuration and member information of the local agent." When you hit this endpoint though, any values that weren't manually set in the user config are returned with values of `0` rather than the default values we assign from the default config. A good example of this is the `raft_snapshot_threshold`. This defaults to `16384` but shows up as `0`. If you manually set it in your config, then it returns the value you set.

#### Reproduction Steps

Steps to reproduce this issue, eg:

1. Start Consul with `consul agent -dev`.
2. Fetch the config from the endpoint with `curl localhost:8500/v1/agent/self | jq`.
3. Look for `RaftSnapshotThreshold`. It will show up as `0`.
4. Now create an HCL config file named `raft.hcl` with `raft_snapshot_threshold = 25000` in it.
5. Stop the agent and restart it with `consul agent -dev -config-file raft.hcl`.
6. Fetch the config again with `curl localhost:8500/v1/agent/self | jq`.
7. You'll see that `RaftSnapshotThreshold` now shows as being `25000`.

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease = 
	revision = 12b16df+
	version = 1.8.4
consul:
	acl = disabled
	bootstrap = false
	known_datacenters = 1
	leader = true
	leader_addr = 127.0.0.1:8300
	server = true
raft:
	applied_index = 19938
	commit_index = 19938
	fsm_pending = 0
	last_contact = 0
	last_log_index = 19938
	last_log_term = 2
	last_snapshot_index = 16384
	last_snapshot_term = 2
	latest_configuration = [{Suffrage:Voter ID:7dc2aeba-3fe4-4288-dd88-6e617229daf2 Address:127.0.0.1:8300}]
	latest_configuration_index = 0
	num_peers = 0
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 2
runtime:
	arch = amd64
	cpu_count = 16
	goroutines = 83
	max_procs = 16
	os = darwin
	version = go1.15.1
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 1
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1

```

</details>

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease = 
	revision = 12b16df+
	version = 1.8.4
consul:
	acl = disabled
	bootstrap = false
	known_datacenters = 1
	leader = true
	leader_addr = 127.0.0.1:8300
	server = true
raft:
	applied_index = 19938
	commit_index = 19938
	fsm_pending = 0
	last_contact = 0
	last_log_index = 19938
	last_log_term = 2
	last_snapshot_index = 16384
	last_snapshot_term = 2
	latest_configuration = [{Suffrage:Voter ID:7dc2aeba-3fe4-4288-dd88-6e617229daf2 Address:127.0.0.1:8300}]
	latest_configuration_index = 0
	num_peers = 0
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 2
runtime:
	arch = amd64
	cpu_count = 16
	goroutines = 83
	max_procs = 16
	os = darwin
	version = go1.15.1
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 1
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1
```

</details>

**Comment by jjshanks** [comment]
I'm interested in doing this.

**Comment by dnephin** [comment]
@jjshanks Thank you, that would be very much appreciated! 

I believe the root of the problem is that these default values are defined in `agent/consul.DefaultConfig()`, but the config that we print in the `/v1/agent/self` endpoint is an `agent/config.RuntimeConfig`.

In `agent/config.DefaultSource()` (where most default values come from) we already call `agent/consul.DefaultConfig()` to get some values, but we aren't using the raft snapshot values yet, they get applied later. If we set the defaults in `agent/config.DefaultSource()` I think it should fix the problem.

**Comment by jkirschner-hashicorp** [comment]
Closed by #8812
